# Copyright: Xulong Wang (xulong.wang@jax.org)
# Purpose: Encode data integration - import the encode data
# Rev: May 15, 2014

source("~/Dropbox/Encode/R/myfunction1.R")
#--- ChIP-seq broadPeak", header = F)) data from Encode -----------------------------------------------------
# setwd("~/Dropbox/Encode/ChIP-seq/marks")
# cbellum.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneCbellumH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# cbellum.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneCbellumH3k27me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# cbellum.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneCbellumH3k4me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# cbellum.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneCbellumH3k4me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# esb4.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneEsb4H3k27acME0C57bl6StdPk.broadPeak", header = F))
# esb4.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneEsb4H3k27me3ME0C57bl6StdPk.broadPeak", header = F))
# esb4.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneEsb4H3k4me1ME0C57bl6StdPk.broadPeak", header = F))
# esb4.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneEsb4H3k4me3ME0C57bl6StdPk.broadPeak", header = F))
# heart.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneHeartH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# heart.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneHeartH3k27me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# heart.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneHeartH3k4me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# heart.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneHeartH3k4me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# kidney.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneKidneyH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# kidney.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneKidneyH3k27me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# kidney.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneKidneyH3k4me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# kidney.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneKidneyH3k4me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# liver.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneLiverH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# liver.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneLiverH3k27me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# liver.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneLiverH3k4me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# liver.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneLiverH3k4me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# mel.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneMelH3k04me1MImmortalC57bl6StdPk.broadPeak", header = F))
# mel.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneMelH3k04me3MImmortalC57bl6StdPk.broadPeak", header = F))
# mel.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneMelH3k27acMImmortalC57bl6StdPk.broadPeak", header = F))
# mel.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneMelH3k27me3MImmortalC57bl6StdPk.broadPeak", header = F))
# smint.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneSmintH3k04me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# smint.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneSmintH3k04me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# smint.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneSmintH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# smint.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneSmintH3k27me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# spleen.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneSpleenH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# spleen.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneSpleenH3k27me3MAdlt8wC57bl6StdPk.broadPeak", header = F))
# spleen.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneSpleenH3k4me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# spleen.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneSpleenH3k4me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# testis.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneTestisH3k04me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# testis.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneTestisH3k04me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# testis.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneTestisH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# testis.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneTestisH3k27me3MAdlt8wC57bl6StdPk.broadPeak", header = F))
# thymus.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneThymusH3k04me1MAdult8wksC57bl6StdPk.broadPeak", header = F))
# thymus.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneThymusH3k04me3MAdult8wksC57bl6StdPk.broadPeak", header = F))
# thymus.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneThymusH3k27acMAdult8wksC57bl6StdPk.broadPeak", header = F))
# thymus.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneThymusH3k27me3MAdlt8wC57bl6StdPk.broadPeak", header = F))
# brain.h3k4me1 <- mycolnames(read.delim("wgEncodeLicrHistoneWbrainH3k04me1UE14halfC57bl6StdPk.broadPeak", header = F))
# brain.h3k4me3 <- mycolnames(read.delim("wgEncodeLicrHistoneWbrainH3k04me3UE14halfC57bl6StdPk.broadPeak", header = F))
# brain.h3k27ac <- mycolnames(read.delim("wgEncodeLicrHistoneWbrainH3k27acUE14halfC57bl6StdPk.broadPeak", header = F))
# brain.h3k27me3 <- mycolnames(read.delim("wgEncodeLicrHistoneWbrainH3k27me3ME14halfC57bl6StdPk.broadPeak", header = F))
# cbellum.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsCbellumCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# cbellum.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsCbellumPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# esb4.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsEsb4CtcfME0C57bl6StdPk.broadPeak", header = F))
# esb4.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsEsb4Pol2ME0C57bl6StdPk.broadPeak", header = F))
# heart.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsHeartCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# heart.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsHeartPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# kidney.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsKidneyCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# kidney.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsKidneyPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# liver.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsLiverCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# liver.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsLiverPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# mel.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsMelCtcfMImmortalC57bl6StdPk.broadPeak", header = F))
# mel.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsMelPol2MImmortalC57bl6StdPk.broadPeak", header = F))
# smint.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsSmintCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# smint.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsSmintPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# spleen.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsSpleenCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# spleen.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsSpleenPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# testis.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsTestisCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# testis.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsTestisPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# thymus.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsThymusCtcfMAdult8wksC57bl6StdPk.broadPeak", header = F))
# thymus.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsThymusPol2MAdult8wksC57bl6StdPk.broadPeak", header = F))
# brain.ctcf <- mycolnames(read.delim("wgEncodeLicrTfbsWbrainCtcfUE14halfC57bl6StdPk.broadPeak", header = F))
# brain.pol2 <- mycolnames(read.delim("wgEncodeLicrTfbsWbrainPol2UE14halfC57bl6StdPk.broadPeak", header = F))
# save.image("~/Dropbox/Encode/R/marks.broadpeak.rdt")
load("~/Dropbox/Encode/R/marks.broadpeak.rdt")
#--- Segment ENCODE broadPeak", header = F)) file -----------------------------------------------------------
samples <- c("cbellum", "esb4", "heart", "kidney", "liver", "mel", "smint", "spleen", "testis", "thymus", "brain")
marks <- c("h3k4me1", "h3k4me3", "h3k27ac", "h3k27me3", "ctcf", "pol2")
n.sps <- length(sample)
n.mks <- length(marks)
g.len <- sum(chrom.length)
marks.bi <- array(data = NA, c(n.sps, n.mks, g.len),  # n.sample, n.mark, 0-1 code 
                  dimnames = list(samples, marks, c(1:sum(chrom.length))))
marks.bi[1, , ] <- cbind(myBinary(cbellum.h3k4me1), myBinary(cbellum.h3k4me3), myBinary(cbellum.h3k27ac), myBinary(cbellum.h3k27me3), myBinary(cbellum.ctcf), myBinary(cbellum.pol2))
marks.bi[2, , ] <- cbind(myBinary(esb4.h3k4me1), myBinary(esb4.h3k4me3), myBinary(esb4.h3k27ac), myBinary(esb4.h3k27me3), myBinary(esb4.ctcf), myBinary(esb4.pol2))
marks.bi[3, , ] <- cbind(myBinary(heart.h3k4me1), myBinary(heart.h3k4me3), myBinary(heart.h3k27ac), myBinary(heart.h3k27me3), myBinary(heart.ctcf), myBinary(heart.pol2))
marks.bi[4, , ] <- cbind(myBinary(kidney.h3k4me1), myBinary(kidney.h3k4me3), myBinary(kidney.h3k27ac), myBinary(kidney.h3k27me3), myBinary(kidney.ctcf), myBinary(kidney.pol2))
marks.bi[5, , ] <- cbind(myBinary(liver.h3k4me1), myBinary(liver.h3k4me3), myBinary(liver.h3k27ac), myBinary(liver.h3k27me3), myBinary(liver.ctcf), myBinary(liver.pol2))
marks.bi[6, , ] <- cbind(myBinary(mel.h3k4me1), myBinary(mel.h3k4me3), myBinary(mel.h3k27ac), myBinary(mel.h3k27me3), myBinary(mel.ctcf), myBinary(mel.pol2))
marks.bi[7, , ] <- cbind(myBinary(smint.h3k4me1), myBinary(smint.h3k4me3), myBinary(smint.h3k27ac), myBinary(smint.h3k27me3), myBinary(smint.ctcf), myBinary(smint.pol2))
marks.bi[8, , ] <- cbind(myBinary(spleen.h3k4me1), myBinary(spleen.h3k4me3), myBinary(spleen.h3k27ac), myBinary(spleen.h3k27me3), myBinary(spleen.ctcf), myBinary(spleen.pol2))
marks.bi[9, , ] <- cbind(myBinary(testis.h3k4me1), myBinary(testis.h3k4me3), myBinary(testis.h3k27ac), myBinary(testis.h3k27me3), myBinary(testis.ctcf), myBinary(testis.pol2))
marks.bi[10, , ] <- cbind(myBinary(thymus.h3k4me1), myBinary(thymus.h3k4me3), myBinary(thymus.h3k27ac), myBinary(thymus.h3k27me3), myBinary(thymus.ctcf), myBinary(thymus.pol2))
marks.bi[11, , ] <- cbind(myBinary(brain.h3k4me1), myBinary(brain.h3k4me3), myBinary(brain.h3k27ac), myBinary(brain.h3k27me3), myBinary(brain.ctcf), myBinary(brain.pol2))
#----------------------------------------------------------------------------------------------
marks.si <- array(data = NA, c(n.sps, n.mks, g.len),  # n.sample, n.mark, signal 
                  dimnames = list(samples, marks, c(1:sum(chrom.length))))
marks.si[1, , ] <- cbind(mySignal(cbellum.h3k4me1), mySignal(cbellum.h3k4me3), mySignal(cbellum.h3k27ac), mySignal(cbellum.h3k27me3), mySignal(cbellum.ctcf), mySignal(cbellum.pol2))
marks.si[2, , ] <- cbind(mySignal(esb4.h3k4me1), mySignal(esb4.h3k4me3), mySignal(esb4.h3k27ac), mySignal(esb4.h3k27me3), mySignal(esb4.ctcf), mySignal(esb4.pol2))
marks.si[3, , ] <- cbind(mySignal(heart.h3k4me1), mySignal(heart.h3k4me3), mySignal(heart.h3k27ac), mySignal(heart.h3k27me3), mySignal(heart.ctcf), mySignal(heart.pol2))
marks.si[4, , ] <- cbind(mySignal(kidney.h3k4me1), mySignal(kidney.h3k4me3), mySignal(kidney.h3k27ac), mySignal(kidney.h3k27me3), mySignal(kidney.ctcf), mySignal(kidney.pol2))
marks.si[5, , ] <- cbind(mySignal(liver.h3k4me1), mySignal(liver.h3k4me3), mySignal(liver.h3k27ac), mySignal(liver.h3k27me3), mySignal(liver.ctcf), mySignal(liver.pol2))
marks.si[6, , ] <- cbind(mySignal(mel.h3k4me1), mySignal(mel.h3k4me3), mySignal(mel.h3k27ac), mySignal(mel.h3k27me3), mySignal(mel.ctcf), mySignal(mel.pol2))
marks.si[7, , ] <- cbind(mySignal(smint.h3k4me1), mySignal(smint.h3k4me3), mySignal(smint.h3k27ac), mySignal(smint.h3k27me3), mySignal(smint.ctcf), mySignal(smint.pol2))
marks.si[8, , ] <- cbind(mySignal(spleen.h3k4me1), mySignal(spleen.h3k4me3), mySignal(spleen.h3k27ac), mySignal(spleen.h3k27me3), mySignal(spleen.ctcf), mySignal(spleen.pol2))
marks.si[9, , ] <- cbind(mySignal(testis.h3k4me1), mySignal(testis.h3k4me3), mySignal(testis.h3k27ac), mySignal(testis.h3k27me3), mySignal(testis.ctcf), mySignal(testis.pol2))
marks.si[10, , ] <- cbind(mySignal(thymus.h3k4me1), mySignal(thymus.h3k4me3), mySignal(thymus.h3k27ac), mySignal(thymus.h3k27me3), mySignal(thymus.ctcf), mySignal(thymus.pol2))
marks.si[11, , ] <- cbind(mySignal(brain.h3k4me1), mySignal(brain.h3k4me3), mySignal(brain.h3k27ac), mySignal(brain.h3k27me3), mySignal(brain.ctcf), mySignal(brain.pol2))
save(marks.bi, marks.si, file = "~/Dropbox/Encode/R/marks.rdt")

#--- RNA-seq ---
name1 <- list.files(path = "~/Dropbox/Encode/RSEM2",
                    pattern = "*.genes.results")
name2 <- gsub(".genes.results", "", name1)
name3 <- unique(gsub("Rep.*", "", name1))  # mouse id
for (i in 1:length(name1)) {
  filepath <- file.path("~/Dropbox/Encode/RSEM2", name1[i])
  cat(i, "/", length(name1), name1[i], "\n")
  assign(name2[i], read.delim(filepath)[, -2])
}

rna1 <- matrix(nrow = nrow(get(name2[1])), ncol = length(name3),  # rna1: TPM 
               dimnames = list(get(name2[1])$gene_id, name3))

for (i in 1:length(name3)) {
  rna1[, i] <- (get(paste(name3[i], "Rep1", sep = ""))$TPM + get(paste(name3[i], "Rep2", sep = ""))$TPM) / 2
}

# heatmap(cor(rna1, method = "pearson"))

rna1 <- as.data.frame(rna1)
ens2sym <- read.delim("~/Dropbox/Info/ensembl2symbol.map", header = F, row.names = 1)
rna1$symbol <- ens2sym[rownames(rna1), ]
rna1 <- aggregate(. ~ symbol, rna1, sum)
rna1 <- data.frame(row.names = rna1$symbol, rna1[, 2:ncol(rna1)])
save(rna1, file = "~/Dropbox/Encode/R/rna1.rdt")

